#CalculateTE.R
#This is an R script that calculates Translation Efficiency (TE) based on RPF.bam, RNAseq.bam and Regions.bed12 inputs
#Objects in this script:
#    GAlignments: The direct input from a bam file to GAlignments object
#    GRanges: An object that stores coordinate information of each transcript, and each exon is a line. This is 1-based
#      All interavals are both ends included: [], which means the intervals: [1, 5], [10, 15], [20, 25]: length: 5+6+6=17
#                  ORF <- GRanges(seqnames = "1",
#                            ranges = IRanges(start = c(1, 10, 20), end = c(5, 15, 25)),
#                            strand = "+")
#    GRangesList: An list object that contains multiple Tx, but you must iterate each Tx as an independent record
#Inputs: 
#    Data.RPF.bam: Converted from a Data.bed13.RPF file, to Data.bed13.RPF.bed1.bed12.sorted.bam
#    Data.RNAseq.bam: Converted from a Data.bed12 file, to Data.bed12.sorted.bam
#    Regions.TE.cds: Annotation in bed12 Format, with cds extracted
#    Regions.TE.cds.len.txt: Generated by CountBed12_Length.sh
#Example: Rscript CalculateTE_BH.R $RPFseq $RNAseq ${OutputName}.TE.cds ${OutputName}.TE.cds.len.txt ${RPFDepth} ${RNADepth} 1 ${OutputName}
#Version: Yu Sun, 2019-07-17
#Version: Yu Sun, 2019-08-05
#	  If you really want to reverse the strand manually, for original stranded RNAseq bam input, do this:
#	  strand(RNAaln) <- ifelse(as.logical(as.character(strand(RNAaln)) == "+"), "-", "+")

args <- commandArgs(TRUE)
InputRPFBam <- args[1]
InputRNABam <- args[2]
AnnotationCDS <- args[3]
LengthFile <- args[4]
RPFDepth <- as.integer(args[5])
RNADepth <- as.integer(args[6])
pseudoCount <- as.integer(args[7])
OutputName <-  args[8]

library(GenomicAlignments)
library(Guitar)
library(ORFik)
print("Extracting BAM from RPF...")
RPFaln <- readGAlignments(InputRPFBam)
print("Extracting BAM from RNAseq...")
RNAaln <- readGAlignments(InputRNABam)

AnnoCDS <-  BED12toGRangesList(AnnotationCDS,header = F)     #This is a GRangesList object
AnnoCDS_Length <- read.table(LengthFile,header = F, col.names = c("Gene", "Length"))
Length <- length(AnnoCDS)
AnnoCDS_tx <- AnnoCDS
names(AnnoCDS) <- paste0("tx",1:Length)
AnnoCDS_PlainNames_1 <- paste(names(AnnoCDS), "_1", sep="")
names(AnnoCDS_tx) <- AnnoCDS_PlainNames_1
print("Calculating Translation efficieny (TE) using ORFik...")
#Use AnnoCDS as tx, rather than the original way: the whole Tx RPKM
final <- translationalEff(grl = AnnoCDS, RNA = RNAaln, RFP = RPFaln, tx = AnnoCDS, with.fpkm = TRUE, pseudoCount = pseudoCount)
final$length <- AnnoCDS_Length$Length
final$CountsRFP <- (final$fpkmRFP-pseudoCount)*final$length*RPFDepth/1000000000
final$CountsRNA <- (final$fpkmRNA-pseudoCount)*final$length*RNADepth/1000000000
final <- final[,c("length","CountsRFP","CountsRNA","fpkmRFP","fpkmRNA","te")]
write.table(final,file=paste(OutputName,".TE.temp.txt",sep = ""),quote=F,col.names = F, row.names = F)
